src/test/scala/sigmastate/utxo/AVLTreeScriptsSpecification.scala

```diff
@@ -1,21 +1,121 @@
 package sigmastate.utxo
 
 import com.google.common.primitives.Longs
+import org.ergoplatform._
+import scorex.crypto.authds.avltree.batch._
 import scorex.crypto.authds.{ADKey, ADValue}
-import scorex.crypto.authds.avltree.batch.{BatchAVLProver, Insert, Lookup}
 import scorex.crypto.hash.{Blake2b256, Digest32}
-import sigmastate._
+import sigmastate.SCollection.SByteArray
 import sigmastate.Values._
+import sigmastate._
 import sigmastate.helpers.{ErgoLikeProvingInterpreter, SigmaTestingCommons}
-import org.ergoplatform._
-import sigmastate.SCollection.SByteArray
 import sigmastate.lang.Terms._
+import sigmastate.serialization.OperationSerializer
 
 class AVLTreeScriptsSpecification extends SigmaTestingCommons {
 
   private val reg1 = ErgoBox.nonMandatoryRegisters.head
   private val reg2 = ErgoBox.nonMandatoryRegisters(1)
 
+  def genKey(str: String): ADKey = ADKey @@ Blake2b256("key: " + str)
+
+  def genValue(str: String): ADValue = ADValue @@ Blake2b256("val: " + str)
+
+  property("avl tree modification") {
+    val prover = new ErgoLikeProvingInterpreter
+    val verifier = new ErgoLikeInterpreter
+    val pubkey = prover.dlogSecrets.head.publicImage
+
+    val avlProver = new BatchAVLProver[Digest32, Blake2b256.type](keyLength = 32, None)
+    val inKey = genKey("init key")
+    avlProver.performOneOperation(Insert(inKey, genValue("init value")))
+    avlProver.generateProof()
+    val digest = avlProver.digest
+    val treeData = new AvlTreeData(digest, 32, None)
+
+    val operations: Seq[Operation] = (0 to 10).map(i => Insert(genKey(i.toString), genValue(i.toString))) :+
+      Update(inKey, genValue("updated value"))
+    val serializer = new OperationSerializer(avlProver.keyLength, avlProver.valueLengthOpt)
+    val opsBytes: Array[Byte] = serializer.serializeSeq(operations)
+    operations.foreach(o => avlProver.performOneOperation(o))
+    val proof = avlProver.generateProof()
+    val endDigest = avlProver.digest
+
+    val prop = EQ(TreeModifications(ExtractRegisterAs[SAvlTree.type](Self, reg1).get,
+      ByteArrayConstant(opsBytes),
+      ByteArrayConstant(proof)), SomeValue(ByteArrayConstant(endDigest)))
+    //    TODO uncomment after fix of "Option constructors are not supported"
+    //    val env = Map("ops" -> opsBytes, "proof" -> proof, "endDigest" -> endDigest)
+    //    val propCompiled = compile(env, """treeModifications(SELF.R4[AvlTree].get, ops, proof) == Some(endDigest)""").asBoolValue
+    //    prop shouldBe propCompiled
+
+    val newBox1 = ErgoBox(10, pubkey)
+    val newBoxes = IndexedSeq(newBox1)
+
+    val spendingTransaction = ErgoLikeTransaction(IndexedSeq(), newBoxes)
```

Comment:
@catena2w @kushti maybe we should require boxesToSpend to be a subset of inputs in transaction? 
I always thought that SELF is just a reference to one of the inputs, for which the script is executing. 
I.e. SELF is always part of INPUTS.
Otherwise to execute all scripts in all inputs we will have to remove corresponding input from the transaction to create a correct context. 
The concrete decision about this should be documented somewhere, if not yet.